rio/src/RIO/File.hs

```diff
@@ -0,0 +1,345 @@
+    ReadMode      -> read_flags
+    WriteMode     -> write_flags
+    ReadWriteMode -> rw_flags
+    AppendMode    -> append_flags
+
+openDir :: MonadIO m => FilePath -> m Fd
+openDir fp
+  -- TODO: Investigate what is the situation with Windows FS in regards to non_blocking
+  -- NOTE: File operations _do not support_ non_blocking on various kernels, more
+  -- info can be found here: https://ghc.haskell.org/trac/ghc/ticket/15153
+ =
+  liftIO $
+  withFilePath fp $ \cFp ->
+    Fd <$>
+    (throwErrnoIfMinus1Retry "openDir" $
+     c_safe_open cFp (ioModeToFlags ReadMode) 0o660)
+
+-- NOTE: Not Async exception safe
+openFileFromDir :: (MonadIO m) => Fd -> FilePath -> IOMode -> m Handle
+openFileFromDir (Fd dirFd) fp iomode =
+  liftIO $
+  withFilePath fp $ \f -> do
+    fileFd <-
+      throwErrnoIfMinus1Retry "openFileFromDir" $
+      c_safe_openat dirFd f (ioModeToFlags iomode) 0o666
+    (fD, fd_type) <-
+      FD.mkFD
+        fileFd
+        iomode
+        Nothing {- no stat -}
+        False {- not a socket -}
+        False {- non_blocking --}
+         -- TODO: We want to investigate if we can do non-blocking here
+       `catchAny` \e -> do
+        void (c_close fileFd)
+        throwIO e
+    -- we want to truncate() if this is an open in WriteMode, but only if the
+    -- target is a RegularFile. ftruncate() fails on special files like
+    -- /dev/null.
+    when (iomode == WriteMode && fd_type == RegularFile) $ Device.setSize fD 0
+    HandleFD.mkHandleFromFD fD fd_type fp iomode False Nothing `onException`
+      (liftIO $ Device.close fD)
+
+-- | Opens a file using the openat C low-level API. This approach allows us to
+-- get a file descriptor for the directory that contains the file, which we can
+-- use later on to fsync the directory with.
+--
+-- @since 0.1.6
+openFileDurable :: MonadIO m => FilePath -> IOMode -> m (Fd, Handle)
+openFileDurable absFp iomode =  do
+  let dir = takeDirectory absFp
+      fp = takeFileName absFp
+
+  dirFd <- openDir dir
+  fileHandle <- openFileFromDir dirFd fp iomode
+  return (dirFd, fileHandle)
+
+-- | This sub-routine does the following tasks:
+--
+-- * It calls fsync and then closes the given handler (mapping to a temporal/backup filepath)
+-- * It calls fsync and then closes the containing directory of the file
+--
+-- These steps guarante that the file changes are durable.
+--
+-- @since 0.1.6
+closeFileDurable :: MonadIO m => Fd -> Handle -> m ()
+closeFileDurable (Fd dirFd) h =
+  liftIO $
+  finally
+    (do fileFd <- handleToFd h
+        fsyncDescriptor "File" (FD.fdFD fileFd) `finally` hClose h
+        fsyncDescriptor "Directory" dirFd)
+    closeDirectory
+  where
+    fsyncDescriptor name fd =
+      void $
+      throwErrnoIfMinus1 ("closeFileDurable - fsync(" <> name <> ")") $
+      c_safe_fsync fd
+    closeDirectory =
+      void $
+      throwErrnoIfMinus1Retry "closeFileDurable - close(Directory)" $
+      c_close dirFd
+
+toTmpFilePath :: FilePath -> FilePath
```

Comment:
I don't think this is sufficient, instead you should use the template file name functions from `System.IO` to ensure the file doesn't already exist.

Reply:
Uhm, you are right, unfortunately, the `System.IO` does not export the functionality to generate only the `FilePath` but rather it opens a `Handle` at the same time, I'll have to either use the existing function and close the file handle or re-implement the file name generation function in this module.